### ISSUE TITLE
Avoid immediately writing assignments and approvals to disk in approval-voting

### ISSUE THREAD

**Issue by rphmeier** [paritytech/polkadot#5317]
in the vein of: https://github.com/paritytech/polkadot/issues/3437

<img width="821" alt="image" src="https://user-images.githubusercontent.com/10121380/163075688-ff67d9b3-6a19-4872-b1e4-0d44c527f90e.png">

We can see empirically that approval-distribution takes around 100-150ms to process a significant proportion of its messages. It processes messages one at a time. Duplicate or irrelevant messages are handled quickly, but new unique messages are sent to the approval-voting subsystem to be imported. This import process blocks on writing to the underlying database. 

We could improve the performance of the approval-distribution gossip subsystem substantially by ensuring that import of assignments and approvals only ever writes to memory, not disk. We can introduce a buffer and a background flushing task. We should ensure that _locally-generated_ assignments and approvals are written to disk, and we should aim to flush to disk on shutdown.

**Comment by rphmeier** [comment]
@sandreim tried out a basic change in #5235 to only flush once every 1000 assignments/approvals are written and it didn't help much. That seems to disprove the original thesis that we were getting backpressured by DB flushes on every message.

So where do these high ToFs come from?

**Comment by sandreim** [comment]
> @sandreim tried out a basic change in #5235 to only flush once every 1000 assignments/approvals are written and it didn't help much. That seems to disprove the original thesis that we were getting backpressured by DB flushes on every message.

The change I tested (https://github.com/paritytech/polkadot/pull/5236/commits/6d29474d4dc67610c168bd1e0e028670920a0a19) improved the db transaction metrics, but had no overall effect on the ToFs.

> So where do these high ToFs come from?

I just checked how many messages are in the `approval distribution` bounded channel, and it looks pretty bad across all validators:
<img width="1375" alt="Screenshot 2022-04-14 at 18 44 23" src="https://user-images.githubusercontent.com/54316454/163426267-09f9d841-9e96-4b15-a4d9-27c3c2d9f80a.png">

**Comment by sandreim** [comment]
I don't really think these tell the full story, as we are only seeing the values at every metric scraping, so anything bursty that goes on in between it is never seen here. It might be even worse...

**Comment by sandreim** [comment]
Similar investigation for dispute-coordinator: https://github.com/paritytech/polkadot/issues/5359

**Comment by alexggh** [comment]
Being an old issue, I did a bit of digging on `versi_v1_10`, it seems the problem persists, this is how `ToF` looks like for the approval-distribution bound queue:


<img width="1257" alt="Screenshot 2023-05-25 at 18 42 31" src="https://github.com/paritytech/polkadot/assets/49718502/d599962f-3710-4ca1-a6cc-9c61173b659d">

I diged a bit into the metrics and indeed it seems there are situations where approval distribution takes longer, because it is waiting after the approval-voting system, see the: `polkadot_parachain_time_awaiting_approval_voting`.

<img width="2119" alt="awaiting_approval_voting_bucket" src="https://github.com/paritytech/polkadot/assets/49718502/c1819059-a1f1-4d82-9c2a-76356f7dec81">

And it seems to be correlated with the `approval-voting` flushing it's db operations, `polkadot_parachain_time_approval_db_transaction`. See how the histogram values look pretty much the same

<img width="2126" alt="polkadot_parachain_time_approval_db_transaction" src="https://github.com/paritytech/polkadot/assets/49718502/40f01f06-77c4-4f19-bc3f-d1809138d694">

So, I think it is worth retrying to move the db flushes out of the main loop and see if there is any impact from that optimization.

**Comment by sandreim** [comment]
Interesting, there are samples above 1s. How many of them are there? (it's hard to tell from the color coding). One difference compared to last investigation is the size of Versi.

**Comment by burdges** [comment]
An mmap supposedly ensures updated wind up written eventually, even if you never flush manually: https://stackoverflow.com/questions/5902629/mmap-msync-and-linux-process-termination

I've heard Windows works similarly, but not sure about macOS.  We've many subsystems which could exploit this given we now assume Linux: assignments, approval votes, grandpa votes, etc.

**Comment by alexggh** [comment]
> Interesting, there are samples above 1s. How many of them are there? (it's hard to tell from the color coding). One difference compared to last investigation is the size of Versi.

There are just a few of them, it might be that there just some single machines that are slower, I'm trying to figure that out from the dashboards.

> An mmap supposedly ensures updates wind up written eventually, even if you never flush manually: https://stackoverflow.com/questions/5902629/mmap-msync-and-linux-process-termination

I haven't diged into it, but that seems to depend on how our database it is implemented. 
I think at first we need to profile where we spend most of our time in approval-voting, flushing is one of the possible culprits, but it might be that even the reads are slowish(and I see a lot of reads happening on the paths where we process a message from approval-distribution subsystem).

**Comment by burdges** [comment]
We do reread records of all approvals announcements and votes for every unapproved candidate every second, well currently every 1/2 second, but doing so should not involve any cryptography, not even hashing, just checking in memory data structures.   in fact, there is no reason for an authenticated db here anyways, so no hashing to maintain a root hash or whatever.

I suppose mmap always requires working directly with serialized structures, like using offsets instead of real pointers or whatever, so yeah that's some complexity.

**Comment by alexggh** [comment]
Looked a bit more closely and it seems we actually go to the database every time for reads when an event is received in approval-voting, the reason for that is, I think, a bug/unintended consequence of our usage of [OverlayedBackend here](https://github.com/paritytech/polkadot/blob/master/node/core/approval-voting/src/lib.rs#L790).

So, because we instantiate OverlayedBackend every time in the main processing loop we will never hit the caching strategy from [here](https://github.com/paritytech/polkadot/blob/master/node/core/approval-voting/src/backend.rs#L132).

**Comment by sandreim** [comment]
Yeah, but likely this is still cached in memory at the DB/page cache level.

**Comment by alexggh** [comment]
Something weird is happening in versi in the past `6h`, I see  some high ToF for `approval-distribution-subsystem`, but without being correlated with high ToF for `approval-voting-subsystem`, any theories why that might happen ?

I would have expected so see high ToF for approval-voting-subsystem as well, if that's what slows it down.

Looking at a single node this is how `approval-distribution` looks like.
<img width="2147" alt="Screenshot 2023-06-02 at 14 10 37" src="https://github.com/paritytech/polkadot/assets/49718502/24374695-e9f3-406e-8d39-db71f80cfb7d">

This is how `approval-voting` looks like.
<img width="2144" alt="Screenshot 2023-06-02 at 14 11 15" src="https://github.com/paritytech/polkadot/assets/49718502/4de4f68b-7016-4d8c-8c9a-a5a4bc09a6a0">

Additionally, looking at the tracing for [import assignment and approval](https://grafana.parity-mgmt.parity.io/goto/ialJQglVR?orgId=1) the time both operation take is in the order  of micro-seconds.

So, I'm starting to think that the slow-down we are seeing in approval-distribution is not triggered by `approval-voting` being slow, but by some other logic subsystem, potentially by `network-bridge-tx-subsystem` when we run `send_assignments_batched` & `send_approvals_batched` 


Just looking at the rate of messages sent to each of the subsystem, `approval-distribution-system` seems to be receiving a steady flow of 3k messages per second, and `approval-voting` around 500, so there is a delta of 2.5k messages which doesn't have to go to approval voting and it might be where the slow-down manifest itself.
<img width="2173" alt="Screenshot 2023-06-02 at 15 33 56" src="https://github.com/paritytech/polkadot/assets/49718502/dd896f14-f6b0-4ede-aef8-5127fdb90f30">

Any thoughts/metrics I could look at to validate these assumptions ?
